### PR TITLE
etcd watcher: centralize error handling

### PR DESCRIPTION
**Pull request by hongchaodeng**
We should centralize error handling in watcher in run(). Otherwise this could silently return.

Also we don't need the grpc code checking anymore. It's fixed.

``` release-note
etcd: watcher centralize error handling
```

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32907)

### CODE REVIEW

**Comment by hongchaodeng** [comment]
cc @timothysc @xiang90

**Comment by hongchaodeng** [comment]
@k8s-bot test this issue: #IGNORE

**Comment by hongchaodeng** [comment]
no previous test result is found...

**Comment by hongchaodeng** [comment]
@k8s-bot unit test this issue: #32770

**Comment by hongchaodeng** [comment]
Can we move forward this?
This could help narrow the problem scope of the watch issue..

**Comment by hongchaodeng** [comment]
@wojtek-t Can you review this?

File: pkg/storage/etcd3/watcher.go [hunk 1]
```diff
@@ -109,6 +107,10 @@ func (wc *watchChan) run() {
 
 	select {
 	case err := <-wc.errChan:
+		if err == context.Canceled {
+			wc.cancel() // just in case
```
Comment (timothysc) [inline pkg/storage/etcd3/watcher.go]: I'm slightly confused why you would cancel the context when it already came back as canceled?

Comment (hongchaodeng) [inline pkg/storage/etcd3/watcher.go]: This is sort of last defense line -- what if we get context.Canceled but `wc.ctx` isn't canceled.
We want to make sure all goroutines in etcd3 watcher to stop regardlessly.

**Comment by wojtek-t** [comment]
LGTM

@hongchaodeng - I'm OOO this week - please don't expect any actions from me this week in general.

**Pull request merged**